### PR TITLE
[feat] File router overhaul

### PR DESCRIPTION
**Pull request by t3dotgg**
Implementation of the new File Router syntax

## TODO
- [x] Unit tests
- [x] Docs
- [x] Handle UI for multiple file types in components better
- [x] Actually enforce file count limit
- [x] Merge infra changes that enable this

## Example

```ts
// Examples
const uploadThingRouter = {
  profilePicture: f(["image"])
    .middleware((req) => auth(req))
    .onUploadComplete((data) => console.log("file", data)),

  messageAttachment: f(["image", "video"])
    .middleware((req) => auth(req))
    .onUploadComplete((data) => console.log("file", data)),

  strictImageAttachment: f({ image: { maxFileSize: "2MB", maxFileCount: 1 } })
    .middleware((req) => auth(req))
    .onUploadComplete((data) => console.log("file", data)),

  mediaPost: f({
    image: { maxFileSize: "2MB", maxFileCount: 4 },
    video: { maxFileSize: "200MB", maxFileCount: 1 },
  })
    .middleware((req) => auth(req))
    .onUploadComplete((data) => console.log("file", data)),
};
```

### CODE REVIEW

File: packages/uploadthing/src/types.ts [hunk 1]
```diff
@@ -25,11 +25,21 @@ type UploadedFile = {
   size: number;
 };
 
-type AllowedFiles = "image" | "video" | "audio" | "blob";
+export type AllowedFileType = "image" | "video" | "audio" | "blob";
 
 export type SizeUnit = "B" | "KB" | "MB" | "GB";
-type PowOf2 = 1 | 2 | 4 | 8 | 16 | 32 | 64 | 128 | 256 | 512 | 1024;
-export type FileSize = `${PowOf2}${SizeUnit}`;
+export type FileSize = `${number}${SizeUnit}`;
```
Comment (juliusmarminge) [inline packages/uploadthing/src/types.ts:31]: I think a generic `number` here kills intellisense and autocompletion

File: packages/react/src/component.tsx [hunk 2]
```diff
@@ -36,7 +43,15 @@ export function UploadButton<TRouter extends void | FileRouter = void>(props: {
...
-  const { maxSize, fileTypes } = permittedFileInfo ?? {};
+  const fileTypes = permittedFileInfo?.config
+    ? Object.keys(permittedFileInfo.config)
+    : [];
+
+  const maxSize = permittedFileInfo?.config
+    ? Object.values(permittedFileInfo.config).map((v) => v.maxFileSize)
+    : [];
+
+  // TODO: Generate better string for end, maybe helper func?
```
Comment (markflorkowski) [inline packages/react/src/component.tsx:54]: Maybe something like

```const formatItem = (key: string, maxFileSize?: string, maxFileCount?: number): string => {
  let item = '';
  if (key === 'blob') {
    item = 'file';
  } else if (key === 'audio') {
    item = 'audio file';
  } else {
    item = key;
  }

  const count = maxFileCount || 1; // default amount
  const size = maxFileSize || '2MB'; //default size

  const pluralSuffix = count > 1 ? 's' : '';
  return `${item}${pluralSuffix} (${size}, max ${count})`;
};

const formattedItems = [];
const { config } = permittedFileInfo

for (const key in config) {
  if (config.hasOwnProperty(key)) {
    const { maxFileSize, maxFileCount } = config[key];
    formattedItems.push(formatItem(key, maxFileSize, maxFileCount));
  }
}

const result = formattedItems.join(', ');```

File: packages/uploadthing/src/internal/handler.ts [hunk 3]
```diff
@@ -197,18 +198,23 @@ export const buildRequestHandler = <
       if (!Array.isArray(files) || !files.every((f) => typeof f === "string"))
         throw new Error("Need file array");
 
+      console.log(
```
Comment (markflorkowski) [inline packages/uploadthing/src/internal/handler.ts:201]: probably should remove/suppress in prod

File: packages/uploadthing/src/upload-builder.ts [hunk 4]
```diff
@@ -4,39 +4,35 @@ import type {
...
-        maxSize: size,
-      });
-    },
-    middleware(resolver) {
-      return createBuilder({
+    middleware(userMiddleware) {
+      return internalCreateBuilder({
         ..._def,
-        middleware: resolver,
+        middleware: userMiddleware,
```
Comment (markflorkowski) [inline packages/uploadthing/src/upload-builder.ts:35]: any reason you renamed this one and not the onUploadComplete() one?

File: packages/uploadthing/src/types.ts (hunk 1 above)
Comment (t3dotgg) [reply packages/uploadthing/src/types.ts:31]: :( cc @markflorkowski

**Comment by t3dotgg** [comment]
Made a new /sink page for testing more stuff, here's the new copy

image

File: docs/src/pages/nextjs/appdir.mdx [hunk 5]
```diff
@@ -24,10 +24,8 @@ const auth = (req: Request) => ({ id: "fakeId" }); // Fake auth function
 // FileRouter for your app, can contain multiple FileRoutes
 export const ourFileRouter = {
   // Define as many FileRoutes as you like, each with a unique routeSlug
-  imageUploader: f
+  imageUploader: f({ image: { maxFileSize: "4MB" } })
     // Set permissions and file types for this FileRoute
```
Comment (markflorkowski) [inline docs/src/pages/nextjs/appdir.mdx:28]: comments should probably be moved above the route def.

File: docs/src/pages/nextjs/pagedir.mdx [hunk 6]
```diff
@@ -41,10 +41,8 @@ const auth = (req: NextApiRequest, res: NextApiResponse) => ({ id: "fakeId" });
 // FileRouter for your app, can contain multiple FileRoutes
 export const ourFileRouter = {
   // Define as many FileRoutes as you like, each with a unique routeSlug
-  imageUploader: f
+  imageUploader: f({ image: { maxFileSize: "4MB" } })
     // Set permissions and file types for this FileRoute
```
Comment (markflorkowski) [inline docs/src/pages/nextjs/pagedir.mdx:45]: same

File: examples/appdir/src/app/sink/page.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import type { OurFileRouter } from "../_uploadthing";
+
+const inter = Inter({ subsets: ["latin"] });
+
+export default function Home() {
+  return (
+    <main className="flex min-h-screen flex-col items-center justify-center  gap-16 p-24">
+      <div className="flex flex-col items-center justify-center gap-4">
+        <div className="flex gap-4">
+          <UploadButton<OurFileRouter>
```
Comment (markflorkowski) [inline examples/appdir/src/app/sink/page.tsx:14]: looks like you lost a title 

```
<span className="text-4xl font-bold text-center">
  {`Upload a file using a button:`}
</span>
```

File: examples/appdir/src/app/sink/page.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            onClientUploadComplete={(res) => {
+              console.log("Files: ", res);
+              alert("Upload Completed");
+            }}
+            onUploadError={(error: Error) => {
+              alert(`ERROR! ${error.message}`);
+            }}
+          />
+
+          <UploadButton<OurFileRouter>
```
Comment (markflorkowski) [inline examples/appdir/src/app/sink/page.tsx:25]: Is there a reason this isn't just in the main `Page.tsx` example?

Also, pagedir and solid examples?

File: examples/pagedir/src/server/uploadthing/router.ts [hunk 9]
```diff
@@ -5,9 +5,7 @@ import { options } from "~/pages/api/auth/[...nextauth]";
 const f = createUploadthing();
 
 export const uploadRouter = {
-  withMdwr: f
-    .fileTypes(["image"])
-    .maxSize("16MB")
+  withMdwr: f(["image"])
```
Comment (markflorkowski) [inline examples/pagedir/src/server/uploadthing/router.ts:8]: comments explaining what this does (default file size, count) would be nice

File: packages/react/src/component.tsx [hunk 10]
```diff
@@ -6,13 +6,62 @@ import {
...
+  const allowedTypes = Object.keys(config) as (keyof ExpandedRouteConfig)[];
+
+  const formattedTypes = allowedTypes.map((f) =>
+    f === "blob" ? "files" : `${f}s`
+  );
+
+  // Format multi-type uploader label as "Supports videos, images and files";
+  if (formattedTypes.length > 1) {
+    const lastType = formattedTypes.pop();
+    return `${formattedTypes.join(", ")} and ${lastType}`;
```
Comment (markflorkowski) [inline packages/react/src/component.tsx:42]: oxford comma? 😂

File: packages/react/src/component.tsx [hunk 11]
```diff
@@ -6,13 +6,62 @@ import {
...
+  const key = allowedTypes[0];
+  const formattedKey = formattedTypes[0];
+
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const { maxFileSize, maxFileCount } = config[key]!;
+
+  if (maxFileCount && maxFileCount > 1) {
+    return `${formattedKey} up to ${maxFileSize}, max ${maxFileCount}`;
+  } else {
+    return `${formattedKey.slice(0, -1)} (${maxFileSize})`;
```
Comment (markflorkowski) [inline packages/react/src/component.tsx:55]: Not a huge fan of pluralizing by default, and slicing off the s if needed, seems kind of backwards?

File: packages/react/src/component.tsx [hunk 12]
```diff
@@ -6,13 +6,62 @@ import {
...
   : keyof TRouter;
 
+const generatePermittedFileTypes = (config?: ExpandedRouteConfig) => {
+  const fileTypes = config ? Object.keys(config) : [];
+
+  const maxFileCount = config
+    ? Object.values(config).map((v) => v.maxFileCount)
+    : [];
+
+  return { fileTypes, multiple: maxFileCount.some((v) => v && v > 1) };
```
Comment (markflorkowski) [inline packages/react/src/component.tsx:23]: if an endpoint accepts multiple types, shouldn't it also always allow multiple files?

File: packages/uploadthing/src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  FileSize,
+} from "./types";
+
+function isRouteArray(
+  routeConfig: FileRouterInputConfig
+): routeConfig is AllowedFileType[] {
+  return Array.isArray(routeConfig);
+}
+
+const getDefaultSizeForType = (fileType: AllowedFileType): FileSize => {
```
Comment (markflorkowski) [inline packages/uploadthing/src/utils.ts:14]: maybe a note in the docs about this?

File: packages/react/src/component.tsx (hunk 12 above)
Comment (t3dotgg) [reply packages/react/src/component.tsx:23]: My gut feel is that someone who didn't input limits would expect either an image OR video, and someone customizing limits would want multiple in one upload

File: packages/uploadthing/src/utils.ts (hunk 13 above)
Comment (t3dotgg) [reply packages/uploadthing/src/utils.ts:14]: Good call

File: packages/react/src/component.tsx (hunk 10 above)
Comment (t3dotgg) [reply packages/react/src/component.tsx:42]: I like me an oxford but this is what copilot gave me

File: examples/appdir/src/app/sink/page.tsx (hunk 8 above)
Comment (t3dotgg) [reply examples/appdir/src/app/sink/page.tsx:25]: I need a sandbox. I want it somewhere in this repo. I don't need more than one sandbox. tRPC has a "kitchen-sink" example. We can do something like that. Or I can just delete this lol.

File: examples/appdir/src/app/sink/page.tsx (hunk 7 above)
Comment (t3dotgg) [reply examples/appdir/src/app/sink/page.tsx:14]: I think you're overestimating how much I care about this file lol

File: docs/src/pages/nextjs/appdir.mdx (hunk 5 above)
Comment (t3dotgg) [reply docs/src/pages/nextjs/appdir.mdx:28]: Will do in docs pass, don't want more conflicts

File: packages/uploadthing/src/internal/handler.ts [hunk 14]
```diff
@@ -24,6 +34,39 @@ export const fileSizeToBytes = (input: string) => {
   return Math.floor(bytes);
 };
 
+const fileCountLimitHit = (
+  files: string[],
+  routeConfig: ExpandedRouteConfig
+) => {
+  const counts = {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/handler.ts:41]:
```suggestion
  const counts: Record<AllowedFileTypes, number> = {
```

to make sure we keep it sync when adding support for more filetypes

**Review by juliusmarminge — COMMENTED** [review]
quick sweep - can take a more thorough review later if you want

File: examples/pagedir/src/server/uploadthing/router.ts (hunk 9 above)
Comment (t3dotgg) [reply examples/pagedir/src/server/uploadthing/router.ts:8]: We should overhaul the examples to be good. After this

**Review by markflorkowski — APPROVED** [review]

**Pull request merged**